packages/protocol/scope-map-cache.ts

```diff
@@ -0,0 +1,24 @@
+import { Location } from "@replayio/protocol";
+
+import { ThreadFront } from "./thread/thread";
+import { defer } from "./utils";
+
+export default class ScopeMapCache {
+  private cache = new Map<string, Promise<Record<string, string>>>();
+
+  async getScopeMap(location: Location): Promise<Record<string, string>> {
+    const cacheKey = `${location.sourceId}:${location.line}:${location.column}`;
+
+    if (this.cache.has(cacheKey)) {
+      return this.cache.get(cacheKey)!;
+    }
+
+    const { promise, resolve } = defer<Record<string, string>>();
+    this.cache.set(cacheKey, promise);
+    const { client } = await import("./socket");
+    const { map } = await client.Debugger.getScopeMap({ location }, ThreadFront.sessionId!);
```

Comment:
We've really gotta harden the frontend around exceptions more. I'd also say this is a good example of why I'm generally against using the `defer` pattern, because if this did throw, the promise will never resolve or reject, it'll just hang forever. 

This function returns a promise, so we should just cache the promise directly, e.g.
```javascript
  async getScopeMap(location: Location): Promise<Record<string, string>> {
    const cacheKey = `${location.sourceId}:${location.line}:${location.column}`;
    let result = this.cache.has(cacheKey);
    if (!result) {
      result = (async () => {
        const { client } = await import("./socket");
        const { map } = await client.Debugger.getScopeMap({ location }, ThreadFront.sessionId!);
        const scopeMap: Record<string, string> = map ? Object.fromEntries(map) : {};
        return scopeMap;
      })();
      this.cache.set(cacheKey, result);
    }
    return result;
  }
```